sockeye/beam_search.py

```diff
@@ -864,6 +880,87 @@ def forward(self,
                             lengths=lengths,
                             estimated_reference_lengths=estimated_reference_lengths)
 
+    def adjust_first_step_masking(self,
+                target_prefix: pt.Tensor, first_step_mask: pt.Tensor) -> pt.Tensor:
+        """
+        Adjust first_step_masking based on the target prefix
+        (Target prefix for each input in the same batch may have a different length. \
+        Thus first_step_mask needs to be adjusted accordingly.)
+
+        :param target_prefix: Shape (batch size, max target prefix length).
+        :param first_step_mask: Shape (batch_size * beam_size, 1)
+        :return (adjusted) first_steps_masking (batch_size * beam_size, max target prefix length + 1).
+
+        An example to illustrate what this function does:
+        Inputs:
+
+        first_steps_masking (batch_size = 2 * beam_size = 5, 1)
+        ([[0],
+        [inf],
+        [inf],
+        [inf],
+        [inf],
+        [0,],
+        [inf],
+        [inf],
+        [inf],
+        [inf]
+
+        and target prefix (batch_size = 2, max target prefix length = 2)
+        tensor([1 2]
+               [1 0])
+        Note: two target prefix tokens in the first sentence, and one target prefix token \
+        in the second sentence)
+
+        Output:
+        Adjusted first_steps_masking (batch_size * beam_size, max target prefix length + 1):
+        tensor([[0, 0, 0],
+                [inf, inf, inf],
+                [inf, inf, inf],
+                [inf, inf, inf],
+                [inf, inf, inf],
+                [0, 0, 0],
+                [inf, inf, 0],
+                [inf, inf, 0],
+                [inf, inf, 0],
+                [inf, inf, 0],
+
+        """
+
+        """
+        Step 1: Create a zero masking matrix with shape (batch size, max target prefix length + 1)
+        Fill 1 into this masking matrix based on the target prefix
+
+        target prefix     initialize masking     masking           roll one step to the right
+                          from target prefix     is not 0           and assign 1 at index 0
+           [1 2]    ->       [1 2 0]     ->      [1 1 0]                   [1 1 1]
```

Comment:
could you explain why we need the roll to the right?

Reply:
Hi Felix, I think it is best to explain it this way: 

If there is no prefix, we add first_step 1 time
If there is a prefix with length 1, we add first step 2 times
If there is a prefix with length 2, we add first step 3 times
If there is a prefix with length 3, we add first step 4 times
...

So the number of times we add first_step is equal to the length of prefix + 1. Rolling to the right reflects this.